command-line/sim_interface.py

```diff
@@ -0,0 +1,171 @@
+#Author : Kevin Litvin
+#Date : November 2023
+#Description : Encapsulates VCU serial communication, Drive communication, sim model generation, and data formatting
+
+from generate_simulation import Simulation, VCU_Pedal, VCU_Pedals
+from UploadDrive import VcuGDriveInterface
+from vcu_simulation import VCUSimulation, ResponseVCU
+from StatesAndEvents import ResponseVCU, EventData, StateData
+from data_generation import DataGeneration
+import time
+import os
+from typing import Union
+
+import json
+
+
+
+
+
+class VCUSimInterface:
+    """
+    The VCUSimInterface class serves as the interface for the VCU Test Framework, providing 
+
+    functionality for simulating the VCU's behavior in race-like scenarios. It encompasses 
+
+    exentensive simulation generation features, VCU serial communication, interfacing 
+
+    with Team Phantom's Google Drive Log Folders, and taylored Data Formatting
+
+    ---------------------------USER DOCUMENT FOR USAGE ------------------------------
+
+    https://docs.google.com/document/d/1h_xk5TsREIRYUv_hB6RO2n2hTzEec8QP7Y_7VT4gCtU/
+
+    ---------------------------------------------------------------------------------
+    """
+    def __init__(self) -> None:
+        """
+        Initialize all components for simulation communication
+
+         Attributes:
+            vcu_writer (VCUSimulation): An instance of the VCUSimulation class for VCU serial communication.
+
+            simulation (Simulation): An instance of the Simulation class for simulation generation
+
+            vcu_gdrive_interface (VcuGDriveInterface): An instance of the VcuGDriveInterface class for interfacing with G Drive Folders
+
+            data_generator (DataGeneration): An reference to the DataGeneration class for data formatting.
+        """
+        self.vcu_writer = VCUSimulation
```

Comment:
A little confused about how this is used. You assign self.vcu_writer as the **literal type VCUSimulation** (not an instance of VCUSimulation) here, but refer to it elsewhere as an instance of VCUSimulation.

Reply:
A reference is passed because we want to have functionality to configure the device if we choose to add more attributes in the constructor. 

These attributes would live in the configs/device_config.json file. When the device is parsed, an instance with the config creates an instance of the writer.

I don't personally like this structure very much either way. I think the config should be parsed within the VCU Class, not the interface. I chose not to touch the VCUCommunication class (previously you had it as VCUSimulation) since it was your initial design. 

I.e I would rather have `self.configure_device` live within the VCUSimulation class, and parsed during the constructor. 

>>> VCUSimulation(config : dict = None) 
>>>      if config = None:
>>>          self.parsejson()
>>>      else:
>>>           self.Find_Port()
>>>           self.baudrate = self.BAUDRATE

Reply:
Ah I see, you're temporarily assigning it the class type and then initializing it later once you have the correct information...**And then** passing it to the Simulation class where it seems like it will never be used by `VCUSimInterface` again. That's a lot of redirection for a simple assignment; up until we parse the JSON, the type annotations are wrong and the ownership/scope of vcu_writer is unclear.

I see why you want  `VCUSimInterface` to have vcu_writer, but having multiple owners of a shared resource is generally a bad idea. Synchronization logic would be spread out across the owners making it harder to manage and maintain. A better design would be a class that protects the shared resource and orchestrates the synchronization. Sort of like the Proxy design pattern. If we decide to test manual control while running a sim model, we'd put the vcu_writer under this proxy.

For now, `VCUSimInterface` is the final class that is composed of all of the components/objects in the test framework. Any internal details of those classes (like vcu_writer) should not be its responsibility. It also seems like a lot of one liner functions. In the future these might get bigger, but we shouldn't prematurely optimize before that.

```
 def __init__(self) -> None:
        """
        Initialize all components for simulation communication including
        Drive Communication, VCU serial writing, Simulation Generation, and Data Formatting
        """
        
        json_config = self._parse_json()
        vcu_writer = device(**json_config)
        
        #Pass in VCUSimulaiton object as reference for thread-lock compatibility.
        #Useful if we decide to test manual control while running a sim model
        self.simulation : Simulation = Simulation(vcu_writer)
        self.vcu_gdrive_interface : VcuGDriveInterface = VcuGDriveInterface()
        self.data_generator = DataGeneration
```

Also, what you refer to as references throughout the code is just passing the class. So data_generator is a class not a reference. I don't think references are a solid concept in Python that you can really control so commenting on it can be confusing.

Reply:
I agree, the structure is a little weird...not sure about putting it in the VCUSimulation class but we can make an issue for any smaller design choices we want to deal with later

Reply:
Both the `VCUSimInterface` and the `Simulation` class use the same instance of a VCU writer. The `Simulation` class is where the bulk of the command-line interface is created, which is where lie the two options: Manual Control, and Simulation Generation.

Manual Control directly uses the simulation writer, where you can send in a 6 command argument:
<APPS1> <APPS2> <BSE> <TSAL> <RTD> <SETRESET>
which directly gets sent to the VCU. 

Simulation Generation takes in the arguments:
<Cycles> <Precision> <APPS Wave> <BSE wave>
Once a simulation is created (potentially consisting of multiple simulation models appended onto one another), the user calls execute. From there, the `VCUSimInterface` is now responsible to send the full simulation to the vcu, refer to the `_write_data(self)` method.

When I say have a thread lock for so we can run manual control while running a simulation. I mean while a simulation is getting executed, the command-line inteface will still allow the user to send in Manual Control commands. We would need a thread because the VCU has a soft-limit of 33ms or else there is no guarantee of data transmission.

You are right, only the Sim Inteface should be responsible for the VCU_Communication instance (the `vcu writer` as we've been referring to). The reason we pass in the instance of a VCU_Communication is because the the command-line interface is defined with the Simulation Class, so it does not have acccess to the any objects in the VCUSimInteface.

The two changes that I now propose instead is this:

1.
The configuration of the `parse_json` and `set_metadata` should reside within the VCU_Communication class, so that that the VCUSimInteface does not need to be responsible for device configuration.

2.
The command line inputs should be defined within the VCUSimInterface. This way, we do not need the VCU_Communciation (vcu_writer) object to be passed to the Simulation class instance in the constructor.
So the user can call Manual Control <APPS1> <APPS2> <BSE> <etc.> and the `VCUSimInterface` will simply call the vcu_writer instance already defined. 

In Conclusion
Ultimately, VCUSimInterface will be the only manager of the vcu writer. It will be responsible for the Manual control and writing the sim model to the VCU. Before the change the Simulation class is responsible for Manual Control since the command line is defined within the Simulation.

Also, there won't be any messy device configs within the interface. The device configs will be done through the constructor of the VCU_Communication (upon initializing the vcu writer)

Reply:
This way, the `Proxy` would also reside within the VCUSimInterface, or we just keep it as a seperate entitiy. Overall, this means managing the writing device would be a lot more simple.

Reply:
## TLRD 

1. The VCU communication class now is responsible for vcu writer device configuration from the Json if no port is passed into the constructor: 50dfb8fbb82cd14a4aee6175028071567087af84

2. The command line inputs are now defined within the Interface class instead of the class that generated simulations. This change no longer means the simulation generator class needs to contain a vcu writer (instance of a `VCU Communication` class) 

0333c7de23e14cdf513d839352bbdbd89d54da86

Reply:
Lastly, I fully agree with you with very long type annotations. That is now fixed

e1c98fe5ebcdd279fccbe6248d0cbb927646c5f8

Reply:
Yup! Agree with the vcu_writer changes. resolving thread